### ISSUE TITLE
[2.2.5] Regression with XML_ResumeParser()

### ISSUE THREAD

**Issue by pointsman** [libexpat/libexpat#204]
With certain XML input XML_ResumeParser() returns XML_STATUS_ERROR although the input is well-formed XML. This is new with expat 2.2.5 and worked correctly up to (and including) 2.2.4.

This program demonstrates the problem:

```
#include <expat.h>
#include <stdio.h>
#include <string.h>

static void
endElement (
    void         *userData,
    const char   *name
    )
{
    if (strcmp(name,"doc") == 0) {
        printf("calling XML_StopParser() in endElement handler\n");
        XML_StopParser((XML_Parser ) userData, 1);
    }
}

static void
printParserState (
    XML_Parser parser
    )
{
    XML_ParsingStatus pstatus;
    
    XML_GetParsingStatus (parser, &pstatus);
    switch (pstatus.parsing) {
    case XML_INITIALIZED:
        fprintf (stderr, "parser status: XML_INITIALIZED\n");
        break;
    case XML_PARSING:
        fprintf (stderr, "parser status: XML_PARSING\n");
        break;
    case XML_FINISHED:
        fprintf (stderr, "parser status: XML_FINISHED\n");
        break;
    case XML_SUSPENDED:
        fprintf (stderr, "parser status: XML_SUSPENDED\n");
        break;
    default:
        fprintf (stderr, "unexpected parser status: %d\n",
                 pstatus.parsing);
        break;
    }
}

int 
main (int argc, char *argv[]) 
{
    XML_Parser parser;
    enum XML_Status result;
    char doc1[] = "<doc/>";
    char doc2[] = "<doc><elem/></doc>";
    char s[255];

    /* This one uses doc1 as input */
    printf ("XML input: '%s'\n", doc1);
    parser = XML_ParserCreate(NULL);
    XML_SetElementHandler(parser, NULL, endElement);
    XML_SetUserData(parser, parser);
    
    result = XML_Parse(parser, doc1, strlen(doc1), 1);
    while (result == XML_STATUS_SUSPENDED) {
        printParserState (parser);
        printf ("... resume parsing\n");
        result = XML_ResumeParser (parser);
    }
        
    switch (result) {
    case XML_STATUS_OK: 
        printf ("XML_STATUS_OK\n"); break;
    case XML_STATUS_SUSPENDED:
        /* Could not happen, would loop above until result !=
         * XML_STATUS_SUSPENDED */
        printf ("unexpected XML_STATUS_SUSPENDED\n"); break;
    case XML_STATUS_ERROR:
        printf ("XML_STATUS_ERROR\n");
        sprintf (s, "%s", XML_ErrorString(XML_GetErrorCode(parser)));
        printf ("\t%s\n", s);
        break;
    }
    printParserState (parser);
    XML_ParserFree(parser);

    /* This one uses doc2 as input */
    printf ("\nXML input: '%s'\n", doc2);
    parser = XML_ParserCreate(NULL);
    XML_SetElementHandler(parser, NULL, endElement);
    XML_SetUserData(parser, parser);
    
    result = XML_Parse(parser, doc2, strlen(doc2), 1);
    while (result == XML_STATUS_SUSPENDED) {
        printParserState (parser);
        printf ("... resume parsing\n");
        result = XML_ResumeParser (parser);
    }
        
    switch (result) {
    case XML_STATUS_OK: 
        printf ("XML_STATUS_OK\n"); break;
    case XML_STATUS_SUSPENDED:
        /* Could not happen, would loop above until result !=
         * XML_STATUS_SUSPENDED */
        printf ("unexpected XML_STATUS_SUSPENDED\n"); break;
    case XML_STATUS_ERROR:
        printf ("XML_STATUS_ERROR\n");
        sprintf (s, "%s", XML_ErrorString(XML_GetErrorCode(parser)));
        printf ("\t%s\n", s);
        break;
    }
    printParserState (parser);
    XML_ParserFree(parser);
    
    return 0;
}
```
Save this into resumeBug.c, build it with expat 2.2.5 (eg. with
`gcc -o resumeBug resumeBug.c -lexpat `
) and run it. I get:
```
XML input: '<doc/>'
calling XML_StopParser() in endElement handler
parser status: XML_SUSPENDED
... resume parsing
XML_STATUS_ERROR
        no element found
parser status: XML_PARSING

XML input: '<doc><elem/></doc>'
calling XML_StopParser() in endElement handler
parser status: XML_SUSPENDED
... resume parsing
XML_STATUS_OK
parser status: XML_FINISHED
```
This is wrong for the first input (&lt;doc/&gt;), this small XML document is well-formed.

The second input shows, that in general it is OK, to call XML_StopParser() with resumable true even in the closing tag of the document element and to call XML_ResumeParser() afterwards.

If I link the test program with expat 2.2.4 (or ealier) and run it I get:
```
XML input: '<doc/>'
calling XML_StopParser() in endElement handler
parser status: XML_SUSPENDED
... resume parsing
XML_STATUS_OK
parser status: XML_FINISHED

XML input: '<doc><elem/></doc>'
calling XML_StopParser() in endElement handler
parser status: XML_SUSPENDED
... resume parsing
XML_STATUS_OK
parser status: XML_FINISHED
```
which is, what I expect.

This is probably related to the changes made to fix issue #137 in 2.2.5.

**Comment by hartwork** [comment]
I don't have time for digging deeper into it right now but thanks for the detailed bug report, that's a great start. Thank you!

**Comment by RMJ10** [comment]
What's happening is that when we exit `XML_Parse()` with nothing left in our input buffer but the `m_processor` field still set to `contentProcessor()`.  `XML_ResumeParser()` then fires off the normal content processor which realises it has no input and errors as it normally would.  The fix is to set `parser->m_processor` to `epilogProcessor()` when we suspend with no input remaining, a quick and simple test.

I'll put together a regression test and a PR for the fix.

**Comment by RMJ10** [comment]
OK, my explanation sucked, let me try again.

When it normally finishes parsing the outermost end tag of a document (or in this case the sole empty tag), the parser switches to using `epilogProcessor` to parse with.  This allows only whitespace and comments through, since those are the only things valid in an XML document once the outermost tag has been closed.

When it suspends at that point, the parser fails to change processors, leaving `contentProcessor` in place.  `contentProcessor` only knows that there are no open tags when it resumes, and supposes that this must mean it is at the start of a document, not the end.  It sees no input, so raises an error as it would if the parser had been fed with an empty document.  Normally `contentProcessor` doesn't need to know the difference between the start and end of the document, because it is swapped out for `epilogProcessor` at the end of the document.

The fix is to swap from `contentProcessor` to `epilogProcessor` before suspending.  That way when the parser resumes, it resumes with `epilogProcessor` and implicitly knows that the document has ended already.

**Comment by hartwork** [comment]
@RMJ10 thanks a lot for looking into it and for your explanation!

Since you said the parser fails to change processors and since this seems to be a regression from (ex 016d08c6f01b0bda286e42ecb0dd77254cc0f071 which I re-shaped back then into) 718d51a66b0d69f3b8c5ff63d044c36dd5314232 as [mentioned here](https://github.com/libexpat/libexpat/pull/138#issuecomment-326768048) I started wondering if there's a way to sort of go back to the past before 718d51a66b0d69f3b8c5ff63d044c36dd5314232 and apply a fixed version that would not introduce the regression, instead.

I would like to propose this patch as an alternative to the current proposed 4a8ae2002a2171fb19f1f5b8652f32594be1fe45 of #205 for review:
```diff
--- a/expat/lib/xmlparse.c
+++ b/expat/lib/xmlparse.c
@@ -2910,9 +2910,11 @@ doContent(XML_Parser parser,
         poolClear(&parser->m_tempPool);
         freeBindings(parser, bindings);
       }
-      if ((parser->m_tagLevel == 0) &&
-          !((parser->m_parsingStatus.parsing == XML_FINISHED) || (parser->m_parsingStatus.parsing == XML_SUSPENDED))) {
-        return epilogProcessor(parser, next, end, nextPtr);
+      if ((parser->m_tagLevel == 0) && (parser->m_parsingStatus.parsing != XML_FINISHED)) {
+        if (parser->m_parsingStatus.parsing == XML_SUSPENDED)
+          parser->m_processor = epilogProcessor;
+        else
+          return epilogProcessor(parser, next, end, nextPtr);
       }
       break;
     case XML_TOK_END_TAG:
```

I think this reverts introduction of the regression and makes the code change processor.
It also seems to pass the test that @pointsman provided (both standalone and intergated like c7860aa4f6d380f73e0e71ab8431f11080a1dc61).
(I ran `make && gcc -o resumeBug resumeBug.c -L./lib/.libs -lexpat && LD_LIBRARY_PATH=lib/.libs/ ./resumeBug`)

@RMJ10 what do you think?

**Comment by RMJ10** [comment]
It looks plausible.  I won't have time to poke it hard for a couple of days, I'm afraid.

**Comment by RMJ10** [comment]
It passes every test I could think of, so yes, let's go with that.  I've updated my PR accordingly.
